Add integral basis and prime factorization for number fields.

skieffer opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Replaces #22168.

#### Brief description of what is fixed or changed

Add functions to compute integral bases and prime factorizations in number fields.

The functions can be accessed directly, or through new methods in the `AlgebraicField` class.


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* polys
  * Add calculation of integral bases for number fields in `polys.numberfields.basis`.
  * Add prime factorization in number fields in `polys.numberfields.primes`.
  * Add other supporting functionality and classes.
  * New functionality can be accessed through new methods in the existing `polys.domains.algebraicfield.AlgebraicField` class.

<!-- END RELEASE NOTES -->

skieffer committed:
Add integral basis and prime factorization for number fields.

* polys
  * Add calculation of integral bases for number fields in `polys.numberfields.basis`.
  * Add prime factorization in number fields in `polys.numberfields.primes`.
  * Add other supporting functionality and classes.
  * New functionality can be accessed through new methods in the existing `polys.domains.algebraicfield.AlgebraicField` class.

skieffer committed:
Code quality: remove unused variable.

skieffer committed:
Include new imports in `__all__` declaration in `sympy.polys`.

skieffer committed:
Adapt for gmpy2.

There were several places in the code that were broken when using gmpy2, so that what had been `int`s became `mpz`s.

skieffer committed:
Adapt docstring examples for gmpy2.

skieffer committed:
Make `DomainMatrix.__truediv__` accept `mpz` denom.

Now we can eliminate some of the conversions to `int` in `polys.numberfields`.

skieffer committed:
Simplify some conversions to `ZZ`.

skieffer committed:
Improve clearing of poly denoms.

We use `dup_clear_denoms()` to stick to lower level operations.

skieffer commented:
By the way, I am writing some doctext for `modules.py` which I had meant to include earlier, to help explain its design plan. I also found one or two small things to fix. (In case you want to hold off in reviewing until I can add these things.)

skieffer committed:
Improve code for mult. tables in numberfield modules.

Refactor and reorganize for symmetry.

skieffer committed:
Improve error messages.

skieffer committed:
Bugfix in `find_min_poly()`.

We were not returning the full list of powers of alpha. (Highest power was omitted.)

skieffer committed:
Support div by arbitrary (nonzero) module elements.

Our other arithmetic ops already convert to the necessary containing module, so no reason not to do the same thing here.
This supports the general understanding that `ModuleElement`s represent algebraic numbers, so all the ordinary arithmetic operations on them should be possible.

skieffer committed:
Improve doctext in `polys.numberfields.modules`.

skieffer committed:
Simplify code in `round_two()` function.

skieffer committed:
Use fully qualified imports.

skieffer committed:
Refactor `add` / `__add__` in `Submodule` class.

Move the type checking to `__add__`. Users who call `add` directly are expected to already know that the operands are compatible.

skieffer committed:
Bring test cov for `polys.numberfields.modules` back to 100%.

jksuom commented:
I have finished reviewing `primes`. It looks good but one problem remains: I could not see where the steps 3 - 5 of Algorithm 6.2.9 are implemented.

skieffer commented:
Quite right -- you do not find them because they are not there! I took the alternative approach described in the remark following the algorithm because (a) I wanted to provide Alg 4.8.17 for valuations at prime ideals anyway, and (b) Cohen says that in practice it is faster.

(I actually think the idea embodied in steps 3-5 is really cool, but still this seemed the better approach.)

I have had other stuff going on, but am hoping to get all the changes discussed so far on this PR made soon. Thanks very much, to both, for your reviewing.

jksuom commented:
> I took the alternative approach described in the remark following the algorithm

I should have noticed that. It seems to work as `Ip` is a product of maximal ideals and therefore `O/Ip` is separable. Only the valuations have to be computed.

I concur with Oscar that this is extraordinary work. I think that it is ready to be merged after some changes that have already been discussed.

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1242 @@
+    """
+    An order in a number field.
+
+    This is an HNF that includes unity.
```
> then it's their fault if they get bad results?

If it's easy to provide the code that can check this then then the docstring here can point to the `is_order()` method and just explain that this will not be checked automatically.

skieffer replied:
Okay, I looked at this one in many different ways. Ultimately, what the functions in `primes` call for is a `Submodule` representing not just an order, but the maximal order, and I don't know of any straightforward way to test for that, aside from essentially running the Round Two algorithm all over again, which would be crazy.

So I've implemented a middle-of-the-road approach:

1. these functions begin by checking some "formal preconditions" for being the right type of `Submodule`, which will at least raise excepetions on obviously ill-formed cases.

2. the boolean properties being checked are pre-set by the `round_two` function, so that, under standard usage, the checks are instantaneous or "free".

Meanwhile, the `Order`, `Ideal`, and `HNF` classes are gone. They weren't needed, so this is a good simplification.

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1242 @@
+            a = QQ(a)
+            if a == 0:
+                raise ZeroDivisionError
+            return self * (1/a)
```
We don't need to explicitly raise ZeroDivisionError here because `1/QQ(0)` already does that.

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1242 @@
+
+
+class ModuleHomomorphism:
+    """A homomorphism from one module to another."""
+
+    def __init__(self, domain, codomain, mapping):
+        self.domain = domain
+        self.codomain = codomain
+        self.mapping = mapping
```
The type and meaning of these parameters needs to be explained. I presume domain and codomain are both of type Module. What should `mapping` be (e.g. a dict)? A doctest example helps a lot in situations like this.

oscarbenjamin replied:
I mapping an arbitrary callable?

skieffer replied:
Yup, it is. Docstring added.

oscarbenjamin commented on sympy/polys/numberfields/primes.py:
```diff
@@ -0,0 +1,490 @@
+        return self.p * self.ZK + self.alpha * self.ZK
+
+    def copy(self):
+        return PrimeIdeal(self.p, self.alpha.copy(), self.e, self.f, self.ZK.copy())
```
There are a few `copy` methods on different classes in this PR. As far as I can tell all of these classes are immutable so it isn't really meaningful to make copies.

skieffer replied:
Agreed. Copying is gone now.

skieffer commented:
Okay, I will be taking all this into account, and hope to get it pulled together over this week and possibly the next.

(Btw, thanks for the kind words @jksuom ! : )

skieffer committed:
Improve `AlgebraicField.integral_basis()`.

Make this return a list of `ANP` instances, instead of a `DomainMatrix` over `QQ`, because:

* This is more in keeping with the overall design of `AlgebraicField` (which uses `ANP` in many places).
* An integral basis is generally thought of as a set (or list is okay) of algebraic numbers (not a matrix).
* The old way was redundant: the matrix over `QQ` can still be obtained easily from `.maximal_order().QQ_matrix`.

skieffer committed:
Improve `__repr__` for `Submodule` and `PowerBasis`.

skieffer committed:
Improving docs... (WIP)

skieffer committed:
Rename `Submodule.container` to `Submodule.parent`.

* Makes the notion of "ancestors" clearer.
* Since we think of the modules as generated over ZZ, and yet their generators may be any QQ-linear combinations of the generators of their parent module, the parent need not strictly be a "container". E.g. an integral basis will be a Submodule of a PowerBasis, but, as generated over ZZ, the former is often larger than the latter.

skieffer committed:
Redesign modular HNF function calls.

* Move the type conversion for modulus `D` (to `ZZ`) up to `sympy.matrices`.
  In `sympy.polys.matrices` it is expected that `D` is already in `ZZ`.
* Do not compute `abs()`; raise `TypeError` if `D` not positive.

skieffer committed:
Improve doctext & comments.

skieffer committed:
Simplify code.

skieffer committed:
Raise `DMDomainError` instead of `TypeError`.

skieffer committed:
Introduce `sympy.polys.numberfields.exceptions` module.

* Special exception classes for the `numberfields` module will be defined here.
* Move `ClosureFailure` from `modules` to here.
* Introduce a new `MissingUnityError` class.

skieffer committed:
Add special case in `DomainMatrix.convert_to()`.

If the domain to convert to is `None`, then simply return a copy.
This is useful in cases where we always want a new matrix, but only sometimes want to change the domain.

skieffer committed:
Improve `PrimeIdeal`, `prime_valuation()`, `_compute_beta()`.

`PrimeIdeal`:
* Rename `as_ideal()` --> `as_submodule()`.
  Easier to understand what this does, plus we will be eliminating `Ideal` class in another commit.
* Do type checking in `__eq__()` method.
* Make the ramification index `e` an optional `__init__` arg, which
  will be computed here if not supplied.
* Reorder `__init__` args, with `ZK` coming first.

`prime_valuation()`:
As a consequence of the changes to `PrimeIdeal`, we are able to now
simplify the possible arguments accepted by this function.

`_compute_beta()`:
* Make the docstring easier to understand.
* Omit the test to discard generators equivalent to rational `p`.
  Users are informed that they should omit it, so we should not
  waste time checking for it.

skieffer committed:
`__eq__()` / `equiv()` for `ModuleElement`

We redesign `ModuleElement.__eq__()` so that a `ModuleElement` can only be equal to another `ModuleElement`.
We add a `ModuleElement.equiv()` method to test all those other cases of equivalence (equality as algebraic numbers) that `__eq__()` used to include.

skieffer committed:
Eliminate unnecessary copying.

The classes `PrimeIdeal`, `Submodule`, and `ModuleElement` are intended to be used as immutables, so there is no need for them to return copies e.g. when being multiplied by 1 or raised to first power.

skieffer committed:
Eliminate the `HNF`, `Order` and `Ideal` classes.

These classes defined in `polys.numberfields.modules` were not really needed, and did not provide actual tests for being an order or an ideal.

We instead give the `Submodule` class some new boolean properties that functions can use to check (at least partial) well-formedness.
The `polys.numberfields.primes` module in particular performs checks on these new properties.

skieffer committed:
Improving docs...(WIP)

skieffer committed:
Repair doctests.

skieffer commented on sympy/polys/domains/algebraicfield.py:
```diff
@@ -362,6 +366,58 @@ def from_GaussianRationalField(K1, a, K0):
+        return ZK.QQ_matrix
+
+    def discriminant(self):
+        """Get the discriminant of the field."""
```
I'm having some trouble with the Sphinx build. Was wondering if you could offer some advice.

I have added some text to the docstring for the `AlgebraicField` class (defined in `sympy.polys.domains.algebraicfield`), which says:

```
    The maximal order, or ring of integers, of the field can also be
    obtained from the :py:meth:`~.maximal_order` method, as a
    :py:class:`~sympy.polys.numberfields.modules.Submodule`.
```
but Sphinx can't seem to resolve the
```
:py:class:`~sympy.polys.numberfields.modules.Submodule`
```
reference at the end. It says,
```
docstring of sympy.polys.domains.algebraicfield.AlgebraicField:171: WARNING: py:class reference target not found: sympy.polys.numberfields.modules.Submodule
```

It's already an absolute path, i.e. starts from `sympy....` so I don't know how I can help it to find this class. Does the class have to be imported into the module where this docstring lives? Because I can't do that, since it causes a circular import.

Have you run into this before? Thanks.

oscarbenjamin replied:
I'm not sure. Maybe just `~.Submodule`?

Easiest way to get help with debugging is to push what you have to the PR.

skieffer committed:
Reformat long line.

skieffer committed:
Include new `numberfields` public functions in global imports.

skieffer committed:
Repair some docstrings.

skieffer committed:
Add some missing docstrings in `numberfields.modules`.

skieffer committed:
Working on docs for `polys.numberfields`... (WIP)

skieffer committed:
Rename `PrimeIdeal.beta` --> `PrimeIdeal.test_factor`.

This is more descriptive of what this quantity actually is.

skieffer committed:
Repair docstring formatting.

skieffer committed:
Improve error message.

skieffer committed:
Handle missing case in `ModuleElement` arithmetic.

Could not divide one `ModuleElement` by another.
(Although a rational number could be divided by a `ModuleElement`.)

skieffer committed:
More docstring improvements for `numberfields`.

skieffer committed:
Improve `numberfields.rst`

* Include selected dunder methods.
* Add a section for `numberfields.utilities`.
* Restructure major section headings ("Internals" vs "Problem Solutions")

skieffer committed:
Move docs for `AlgebraicNumber`.

Move this to the docs page for `sympy.core.numbers`, since that is where this class is actually defined.

The docs were the last remaining reason why we needed to import `AlgebraicNumber`
into `polys.numberfields`, so we can now cut this import.

skieffer committed:
Improve `numberfields` docs.

* Rename section heading ("Modules" --> "Finitely-generated modules")
* Fix some typos, etc.

skieffer committed:
Prune unused import.

skieffer committed:
Add matrix normal forms to docs.

Now included under both `sympy.matrices` and `sympy.polys.matrices`.

skieffer commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1242 @@
+
+    @property
+    def container(self):
+        """A larger module, in which this one is regarded as a submodule."""
```
Agreed, and the maximal order itself is a good example, often being larger than the power basis, in the sense of Z-span. Calling it `parent` also makes the notion of "ancestor" more obvious.

skieffer commented:
Okay, I think I have addressed all the points discussed above, including adding `polys.numberfields` to the docs.

Sorry there are so many commits. Most are just working on the docs. Only a few commits change code, and the messages should make this clear. 

I've tried to summarize the changes, outside of improving the docs:

* Eliminate the HNF, Order, and Ideal classes
    - Introduce new function to help test formal properties of Submodules
* Introduce `polys.numberfields.exceptions` module
* Revise `ModuleElement.__eq__()`, add `ModuleElement.equiv()`
* Improve division for `ModuleElement`
* Rename `Submodule.container` --> `Submodule.parent`
* Redesign `AlgebraicField.integral_basis()`: now it returns a list of `ANP` instances
* Rename `PrimeIdeal.beta` --> `PrimeIdeal.test_factor`
* Rename `PrimeIdeal.as_ideal` --> `PrimeIdeal.as_submodule`
* Eliminate superfluous copying -- classes are immutable, so no need
* Move `int` conversion from Hermite normal form function in `polys` into corresp. func in `sympy.matrices`
* Remove unnecessary import of `AlgebraicNumber` into `polys.numberfields`
* Build special case into `DomainMatrix.convert_to` (passing `None` just gets a copy)
* Simplify some code
* Use fully qualified imports

skieffer committed:
Use fully qualified imports.

skieffer committed:
Remove conflict with master branch.

skieffer committed:
Prune unused import.

skieffer committed:
Supply missing import.

skieffer committed:
Make docstrings raw.

skieffer committed:
Patch unit test for with/without gmpy2.

skieffer committed:
Touch up `numberfields` docstrings.

skieffer commented:
I spotted a couple of things to fix up in the docs, which I've just pushed.

Okay, I think I'm done for now.

oscarbenjamin commented on doc/src/modules/polys/numberfields.rst:
```diff
@@ -0,0 +1,289 @@
+
+At present only a subset of the tasks enumerated in the lists above is yet
+supported in SymPy, and if you are interested in expanding support, you are
+encouraged to contribute!
```
This note here is an important one. I think it would be good to make clear that everything needed is in Cohen (if that is the case).

Am I right in thinking that you yourself intend to build on this by implementing some of tasks 4-9?

skieffer replied:
Yes, I would like to implement tasks 4-9, and/or work with others who might be keen on it. And yes, I believe it's all in Cohen. Will add note.

oscarbenjamin commented on doc/src/modules/polys/numberfields.rst:
```diff
@@ -0,0 +1,289 @@
+
+Algebraic number fields are represented in SymPy by the
+:py:class:`~.AlgebraicField` class, which is a part of the
+polynomial domains system.
```
I think it would be good to add a
```
:ref:`polys-domainsref`
```
here which links through to here:
https://docs.sympy.org/latest/modules/polys/domainsref.html#reference-docs-for-the-poly-domains

skieffer replied:
Ah, good. I didn't know how to make links of this kind (i.e. using `:ref:`).

oscarbenjamin commented on doc/src/modules/polys/numberfields.rst:
```diff
@@ -0,0 +1,289 @@
+
+In SymPy, there is a distinction between number and expression classes defined
+in the :py:mod:`sympy.core.numbers` module on the one hand, and domains and
+domain elements defined in the :py:mod:`~sympy.polys` module on the other.
```
This is explained in more detail in
```
:ref:`polys-domainsintro`
```
which links to here:
https://docs.sympy.org/latest/modules/polys/domainsintro.html

oscarbenjamin commented on doc/src/modules/polys/numberfields.rst:
```diff
@@ -0,0 +1,289 @@
+conjugacy; in other words, the number and all its conjugates are algebraically
+indistinguishable.
+
+In order to select a particular conjugate, we can identify
+it with a point in the complex plane. This means identifying two real
+intervals, one on the real axis and one on the imaginary axis, such that
+exactly one root of the given irreducible polynomial over $\mathbb{Q}$ lies
+within the complex region so defined.
+
+Support for this in SymPy is currently limited. The :py:func:`~.isolate`
+function is currently applicable only to algebraic numbers lying on the
+real line.
```
How exactly is this different from RootOf?
```python
In [8]: rs = [RootOf(x**5 - x + 1, n) for n in range(5)]

In [9]: [r.n(4) for r in rs]
Out[9]: [-1.167, -0.1812 - 1.084⋅ⅈ, -0.1812 + 1.084⋅ⅈ, 0.7649 - 0.3525⋅ⅈ, 0.7649 + 0.3525⋅ⅈ]

In [10]: rs[0]
Out[10]: 
       ⎛ 5           ⎞
CRootOf⎝x  - x + 1, 0⎠

In [11]: rs[0].is_real
Out[11]: True

In [12]: rs[1].is_real
Out[12]: False

In [13]: rs[1].conjugate() == rs[2]
Out[13]: True
```

skieffer replied:
Interesting. TBH, on the code that I didn't write (anything in `numberfields.(minpoly|numbers|isomorphism)` I was doing my best to comment, without knowing the full story. I'll look into this some more.

oscarbenjamin replied:
The code underlying RootOf is in `sympy/polys/rootisolation.py` e.g. `dup_isolate_all_roots`:
```python
In [1]: from sympy.polys.rootisolation import dup_isolate_all_roots

In [2]: dup_isolate_all_roots([ZZ(1), ZZ(0), ZZ(2)], ZZ)
Out[2]: 
([],
 [(((mpq(0,1), mpq(-4,1)), (mpq(4,1), mpq(0,1))), 1),
  (((mpq(0,1), mpq(0,1)), (mpq(4,1), mpq(4,1))), 1)])
```
This shows that `x**2 + 2` has no real roots and two complex roots of multiplicity 1 which are bounded by `(0, 4) + (-4, 0)*I` and `(0, 4) + (0, 4)*I`

skieffer replied:
I have looked more closely at this funny `isolate()` function under `numberfields`, and here's what I find:

It is not used at all in SymPy (outside of the one unit test devoted to it), but it is
included in `from sympy import *`.

My best guess is that it exists to support something involving NumPy, SciPy, or other.
This is suggested by the use of `lambdify()` in its implementation. (Docstring for `lambdify` talks
about making SymPy objects usable by NumPy etc.)


Q: How does it compare to `RootOf` / `dup_isolate_all_roots`?

A: It _uses_ such things internally, and its purpose is to accept an algebraic number (not a poly),
_find_ a poly for that number, and then do some isolation.


Q: Why the restriction to _real_ algebraic numbers, since SymPy clearly does support isolating boxes in `CC`?

A: Not clear. Probably pertains to the intended external application, involving NumPy or whatever it was.


Having now reviewed this, I'd like to (a) improve the docstring, and (b) move the function. Please see my
latest comment on this PR.

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1794 @@
+            Hermite Normal Form.
+        hnf_modulus : :ref:`ZZ`, None, optional
+            If a positive integer is provided, use this as modulus in the
+            HNF reduction.
```
This is explained in the `hermite_normal_form` function docstring so a clickable link from here to there like "see the explanation of this parameter at :py:func:..." would be good or at least a "See also" here.

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1794 @@
+        """
+        return None
+
+    def represent(self, elt):
+        """
+        Represent an element of an ancestor module as an integer-linear
+        combination over the generators of this module.
```
Maybe this docstring should be the one that actually explains what `represent` means rather than the two subclasses. I think it would be good to have (somewhere) a demonstration of how represent works showing how a particular number is represented in a particular module and how you could reconstruct the `Expr` form of a number from the generators and the vector returned here by represent.

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1794 @@
+        Explanation
+        ===========
+
+        When *other* is another :py:class:`~.ModuleElement`, the test for
+        equivalence requires not just that the two elements represent the same
+        algebraic number, but that they have the same :py:class:`~.PowerBasis`
+        ancestor.
```
It seems odd to me that it's okay for other to be something completely different that isn't even a ModuleElement but if it is a ModuleElement then it needs to have exactly the same ancester. Why is it useful to have a method that works like this?

skieffer replied:
Well, it was restricted to these cases only because they are the easy ones (and the only ones I needed!).

Without a common `PowerBasis` ancestor, it essentially means asking whether an element of `QQ(alpha)` and an element of `QQ(beta)` are equal, with `alpha` and `beta` being algebraic numbers not necessarily bearing any obvious relation to one another.

I guess it wouldn't hurt to handle this case too, using the `to_number_field` function. It just means bringing in some heavy machinery that's not required for the cases currently handled.

skieffer replied:
On second thought, I think the remaining case (`ModuleElement`s lacking a common ancestor) has to either wait, or just be dropped. It's really not needed at all, and it was just going to be a "bonus" for "completeness".

I think it's perfectly fine to raise `UnificationFailed` here. Also I now make it clear in the docstring that this is a deliberate limitation.

The issue is, I am actually having trouble getting `minpoly.py` and `subfield.py` to answer this question, in the form in which it's given. I'm not yet able to say whether they actually _can't_ do it, or I simply haven't figured out how. In any case, it doesn't seem to be easy.

If it turns out to be easy, I will come back and supply this case. If it turns out to be hard, then we might be talking about some sizable new developments in `minpoly.py` and/or `subfield.py`, which would have to be done in a  separate PR.

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1794 @@
+
+    def __mod__(self, a):
+        r"""
+        Reducing a :py:class:`~.ModuleElement` mod an integer *a* reduces all
+        numerator coeffs mod $d a$, where $d$ is the denominator of the
+        :py:class:`~.ModuleElement`.
+
+        Explanation
+        ===========
+
+        For example, if this :py:class:`~.ModuleElement` represents the number
+
+        $\frac{15 a_1 + a_0}{2} = \frac{a_1 + a_0}{2} + 7 a_1$
+
+        then reducing mod 7 should mean throwing away that part that is a poly
+        in the basis elements $a_i$, with content divisible by 7. But this is
+        achieved by reducing our coeffs mod $2 \times 7$.
```
I'm finding it hard to get my head round what this means.

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1794 @@
+    def make_compat(self, other):
+        """
+        Try to make a compatible pair of :py:class:`~.ModuleElement`, one
+        equivalent to this one, and one equivalent to the other.
```
In other contexts in the polys domains this is referred to as unification and the method would be called `unify`. Maybe it would make sense to use that name here or maybe it would be confusing...

oscarbenjamin commented on sympy/polys/numberfields/modules.py:
```diff
@@ -0,0 +1,1794 @@
+
+    def represent(self, element):
+        r"""
+        Represent an element of this endomorphism ring, as a single column
+        vector.
+
+        Explanation
+        ===========
+
+        Let $M$ be a module, and $E$ its ring of endomorphisms. Let $N$ be
+        another module, and consider a homomorphism $\varphi: N \rightarrow E$.
+        In the event that $\varphi$ is to be represented by a matrix $A$, each
+        column of $A$ must represent an element of $E$. This is possible when
+        the elements of $E$ are themselves representable as matrices, by
+        stacking the columns of such a matrix into a single column.
+
+        This method supports calculating such matrices $A$, by representing
+        an element of this endomorphism ring first as a matrix, and then
+        stacking that matrix's columns into a single column.
```
Is it possible to have a simple example here?

oscarbenjamin commented on doc/src/modules/polys/numberfields.rst:
```diff
@@ -0,0 +1,289 @@
+
+
+Solving the Main Problems
+=========================
```
Reading this in the rendered docs this title seems odd coming after the reference docs for the classes and functions above. I wonder if this belongs further up.

oscarbenjamin commented on doc/src/modules/polys/numberfields.rst:
```diff
@@ -0,0 +1,289 @@
+The Primitive Element Problem
+-----------------------------
+
+.. autofunction:: primitive_element
```
This function is badly under-documented. When I wrote the other docs for domains this one slipped under the radar.

oscarbenjamin commented on doc/src/modules/polys/numberfields.rst:
```diff
@@ -0,0 +1,289 @@
+.. autofunction:: primitive_element
+
+
+The Field Isomorphism Problem
+-----------------------------
+
+.. autofunction:: field_isomorphism
+
+
+The Field Membership Problem
+----------------------------
+
+.. autofunction:: to_number_field
```
The docstrings here are lacking.

skieffer replied:
I'll give it a shot. (Again, this is not my code, but I should probably be getting familiar enough with these to write something reasonable!)

oscarbenjamin replied:
Sorry I thought that you had added these functions.

skieffer replied:
No worries, I've been meaning to familiarize myself with these anyway.

skieffer committed:
Misc. improvements in `polys.numberfields.primes`

* Give the pretty printing method better keyword args, and better documentation.
* `PrimeIdeal.as_submodule()` method: better docs, and make it automatically set boolean properties in the produced `Submodule`.
* Supply missing `PrimeIdeal.__add__()` method. (Adding ideals yields their GCD.)
* Provide docs for `PrimeIdeal.__add__()` and `PrimeIdeal.__mul__()`.
* Improve other docstrings.

skieffer commented:
I've now reviewed all the public functions that existed under `numberfields` before I started doing any work
on SymPy, and this has resulted in a few things:

* I have written expanded docstrings for all these functions, and improved the test coverage.
* I have realized that the way I split it up before (into three modules, `minpoly.py`, `isomorphism.py`, and `numbers.py`) is really not the best way.

I've opened a PR, #22464, that

* reorganizes
* adds docs
* improves test coverage

and I'm wondering what you think. In particular, what would you think about merging that one, and then I could rebase this one on it.

skieffer committed:
Add `PrimeIdeal.reduce_poly()` method.

oscarbenjamin commented:
> and I'm wondering what you think. In particular, what would you think about merging that one, and then I could rebase this one on it.

That sounds reasonable. I'll focus on that one first.

skieffer committed:
Improve docstrings in `poly.numberfields.modules`.

* Reorganize docstrings for the three `represent` methods under superclass `Module`.
* Add references for HNF modulo D argument.
* Clarify `ModuleElement.__mod__`.
* Add example for `EndomorphismRing.represent`.

skieffer committed:
Rename `ModuleElement.make_compat()` --> `unify()`.

skieffer committed:
Merge branch 'master' into algnt_ext2

skieffer committed:
Improve docstring for `polys.numberfields.subfield`.

skieffer committed:
Improve use of unification in `polys.numberfields.modules`.

* We now raise `UnificationFailed`, instead of returning `(None, None)` when we cannot unify.
* `ModuleElement.equiv()` now raises `UnificationFailed` when appropriate.

skieffer committed:
Fix up docstrings for `polys.numberfields`.

skieffer committed:
Prune unused import.

skieffer commented:
I think I've now addressed all the points raised above.

Besides that, I gave the `PrimeIdeal` class a `reduce_poly()` method.
In the commit before that, I also made some minor improvements in `PrimeIdeal`.
The commit messages explain.

skieffer committed:
Repair doctest in `AlgebraicField`.

oscarbenjamin commented:
@jksuom I haven't yet gone through this again but does it look good to you?

jksuom commented:
This looks fine to me, ready to be merged.

oscarbenjamin commented:
Looks good, thanks!

Thanks for the review @jksuom.

@skieffer this is fantastic. I'm looking forward to what comes later!

skieffer commented:
Thank you @oscarbenjamin and @jksuom for reviewing! It's a ton of work to review this much code, and I truly appreciate it. Your comments resulted in many substantial improvements. I am definitely proud of this contribution to SymPy! And yes, looking forward to more!
